Example and dev doc for new rotate (#6073)
sunpy/sunpy: docs/dev_guide/contents/map_rotate_custom.rst

.. _map_rotate_custom:

****************************
Adding a new rotation method
****************************

> It is possible to select from a number of rotation methods when using :func:`sunpy.image.transform.affine_transform` and :meth:sunpy.map.GenericMap.rotate`.

Reviewer: ```suggestion
It is possible to select from a number of rotation methods when using :func:`sunpy.image.transform.affine_transform` and :meth:`sunpy.map.GenericMap.rotate`.
```

Author: I wonder why that didnt raise a warning I could find in the doc build.

Reviewer: Heh, I wondered the same thing.  I guess it was so confused by it that it simply interpreted it as straight text.

.. _map_rotate_custom:

****************************
Adding a new rotation method
****************************

If you want to add a new rotation method that can be used by `sunpy.map.GenericMap.rotate`, you can do the following:

.. code-block:: python

    from sunpy.image.transform import add_rotation_function

    @add_rotation_function("my_rotate", handles_clipping=False, handles_image_nans=False, handles_nan_missing=False)
    def _rotation_mine(image, matrix, shift, order, missing, clip):
        # Rotation code goes here
        return rotated_image

The best explanation of the arguments is found from the detailed function documentation:

.. autofunction:: sunpy.image.transform.add_rotation_function
   :noindex:

Then you can do:

.. code-block:: python

    from sunpy.map import Map
    from sunpy.data import sample

    hmi_map = sunpy.map.Map(sample.HMI_LOS_IMAGE)

    rot_map = hmi_map.rotate(order=3, recenter=True, method="my_rotate")

> It is important to know what keywords from `sunpy.map.GenericMap.rotate` are used by the underlying rotation function.
> 
> * ``angle`` - This is passed to `sunpy.image.transform.affine_transform` as a rotation matrix.
> * ``rmatrix`` - This is passed to `sunpy.image.transform.affine_transform`
> * ``order`` - This is passed to `sunpy.image.transform.affine_transform`
> * ``scale`` - This is passed to `sunpy.image.transform.affine_transform`
> * ``recenter`` - This is passed to `sunpy.image.transform.affine_transform`
> * ``missing`` - This is passed to `sunpy.image.transform.affine_transform`
> * ``use_scipy`` - This is passed to `sunpy.image.transform.affine_transform`, it is deprecated and removal is planned for sunpy 4.1, you should use ``method`` instead.
> * ``method`` - This is passed to `sunpy.image.transform.affine_transform`
> * ``clip`` - This is passed to `sunpy.image.transform.affine_transform`

Reviewer: ```suggestion
```
How is this "important to know"?  I presume you added this in response to @dstansby's comment (https://github.com/sunpy/sunpy/pull/6073/files#r851439078), but I don't understand the value in this description.  Why would a developer care how parameters are passed along?

> The methods that ``sunpy`` use are defined in `sunpy/image/transform.py <https://github.com/sunpy/sunpy/blob/main/sunpy/image/transform.py>`__`.

Reviewer: ```suggestion
The available rotation methods are all implemented using the :func:`~sunpy.image.transform.add_rotation_function` decorator, so you can look in `sunpy/image/transform.py <https://github.com/sunpy/sunpy/blob/main/sunpy/image/transform.py>`__ for examples of how to use this decorator.
```

.. _map_rotate_custom:

****************************
Adding a new rotation method
****************************

If you want to add a new rotation method that can be used by `sunpy.map.GenericMap.rotate`, you can do the following:

.. code-block:: python

    from sunpy.image.transform import add_rotation_function

    @add_rotation_function("my_rotate", handles_clipping=False, handles_image_nans=False, handles_nan_missing=False)
    def _rotation_mine(image, matrix, shift, order, missing, clip):
        # Rotation code goes here
        return rotated_image

The best explanation of what the arguments mean is:

.. autofunction:: sunpy.image.transform.add_rotation_function
   :noindex:

> Then you can do:

Reviewer 2: Can you add a sentence that explains how keyword arguments are passed from `rotate()` to the custom rotate function?

Author: Fixed maybe?

.. code-block:: python

    from sunpy.map import Map
    from sunpy.data import sample

    hmi_map = sunpy.map.Map(sample.HMI_LOS_IMAGE)

    rot_map = hmi_map.rotate(order=3, recenter=True, method="my_rotate")

> Below is a non-functioning OpenCV2 rotation method as an example:

Reviewer 2: Why non-funcitoning? Can we make it functioning without much effort?

Author: It lacks proper handling of the shift (based off the comments from the original PR) and someone who understands the differences in the affine transform in openCV2 could make it functioning with some effort.

.. _map_rotate_custom:

****************************
Adding a new rotation method
****************************

If you want to add a new rotation method that can be used by `sunpy.map.GenericMap.rotate`, you can do the following:

.. code-block:: python

    from sunpy.image.transform import add_rotation_function

    @add_rotation_function("my_rotate",

>                         handles_clipping=False, handles_image_nans=False, handles_nan_missing=False)

Reviewer 3: Is the indentation really out on these lines like it renders on GH or is GH having a fit?

Author: That looks like a rendering on the diff.

    def _rotation_mine(image, matrix, shift, order, missing, clip):
        # Rotation code goes here
        return rotated_image

Then you can do:

.. code-block:: python

    from sunpy.map import Map
    from sunpy.data import sample

    hmi_map = sunpy.map.Map(sample.HMI_LOS_IMAGE)

    rot_map = hmi_map.rotate(order=3, recenter=True, method="my_rotate")

There are several things you will want to be aware of when adding a new rotation method.

> * The `.add_rotation_function` means you can handle nans, clipping, and missing values in the image array

Reviewer 3: "you can handle" here is ambiguous, does the function have to handle it or the decorator. The decorator isn't "me".

I think we need to describe the expected inputs and outputs of a rotation function first. Things like you shouldn't modify the input array, the function + decorator must handles NaNs, handle missing values, clipping etc. Then after the expectations of the function are setup then we can say "and there is this decorator which can do these things for you".

Basically I am saying drop the list and describe the function, then introduce the details of the decorator as an aid.

Author: How about I just replicate the very detailed doc string that Albert wrote?

Author: Done.

  before it is passed to your rotation function via the boolean flags.
  The documentation in `.add_rotation_function` explain how these work.
* The arguments into the rotation method must be ``image, matrix, shift, order, missing, clip``.
  There is no variation allowed.
* The return value of the rotation method must the rotated image array.
* You must describe the rotation method in the docstring of the function using unnumbered bullet points.

> Below is a semi-functioning OpenCV2 rotation method as an example:

Reviewer 3: You probably need to elaborate on semi-functioning or people are going to ask why this isn't just in the code.
